### PR TITLE
test: harden the conftest fixtures and close the IPv6 and SASL gaps

### LINKED ISSUE #274: Harden test/conftest.py and close the IPv6 and SASL test gaps
Part of the repository modernization effort.

## Background

`test/conftest.py` starts real `memcached` processes. The fixtures are not robust.

### Problem 1: no start check

`test/test_tls.py:39-40` checks `p.poll()` and calls `pytest.skip(...)`. The four fixtures in `test/conftest.py` do no such check.

If `memcached` is absent from `PATH`, every test fails with an opaque `ConnectionRefusedError`. A clear skip message is better.

### Problem 2: fixed ports

The fixtures use port `11211`, port `5000` (`test/conftest.py:27`), and port `5001` (`test/test_tls.py:25`). No collision check runs. Two runners on one host break each other. A leftover `memcached` process breaks the run.

### Problem 3: fixed sleep

The fixtures call `time.sleep(0.1)` after start. A slow start causes a flaky failure.

Fix: poll the port or the socket until it accepts a connection. Use a timeout.

### Problem 4: deprecated pytest API

`test/conftest.py:12,23,33,45` use `pytest.yield_fixture`. This is a deprecated alias of `@pytest.fixture`.

### Problem 5: stale unix socket

The unix socket path is `/tmp/memcached.sock`. See `test/test_socket.py:11`. If `p.kill()` leaves the socket file, the next run fails to bind.

### IPv6 has no end-to-end test

`test/conftest.py:44-56` starts a real server with `-l::1`. No test connects to it.

The only `::1` references live in `test/test_server_parsing.py:30-60`. Those tests call `Protocol.split_host_port()`. They parse strings. They never open a socket.

The fixture runs on every session and tests nothing.

Fix: add a real IPv6 round trip test, or delete the fixture. Do not keep an unused fixture.

### SASL auth has no integration test

`test/test_auth.py:16-61` mocks `Protocol._get_response` and feeds canned bytes. It tests the client state machine. It never starts a `memcached` with auth enabled.

No fixture starts an auth-enabled server.

Fix: add a fixture that starts `memcached` with SASL enabled. Add one test that authenticates and does a set and a get.

### Distributed hashing is thin

`test/test_distributed_client_hashing.py` holds 12 lines. It asserts that one key maps to one server, ten times.

It never tests ring rebalance. That is the interesting property of consistent hashing.

Fix: add a test that adds or removes a server and checks the key-to-server mapping.

## Coverage that is adequate

- TLS: `test/test_tls.py` covers it. Run `34284112624` ran 48 of 48 TLS tests with 0 skips.
- Compression: `test/test_compression.py` covers zlib, a bz2 swap, a mismatch, and the mocked toggles.
- Unix socket: `test/test_socket.py` reuses `MemcachedTests`. It holds no socket-specific edge cases. This is thin but acceptable.

## Acceptance criteria

- [ ] Every fixture in `test/conftest.py` checks `p.poll()` after start. It calls `pytest.skip()` with a clear message when the process died or `memcached` is absent from `PATH`.
- [ ] Every fixture replaces `time.sleep(0.1)` with a connect-retry loop and a timeout.
- [ ] `pytest.yield_fixture` is replaced with `@pytest.fixture` at `test/conftest.py:12,23,33,45`.
- [ ] The unix socket file is removed on teardown.
- [ ] The IPv6 fixture has a test that does a set and a get over IPv6, or the fixture is deleted.
- [ ] A SASL-enabled `memcached` fixture exists. One test authenticates against it and does a set and a get.
- [ ] `test/test_distributed_client_hashing.py` tests the key-to-server mapping after a server changes.
- [ ] The suite passes locally and in CI.

## Files to change

`test/conftest.py`, `test/test_distributed_client_hashing.py`, `test/test_auth.py`, `test/test_server_parsing.py`. New files are acceptable, for example `test/test_ipv6.py` and `test/test_sasl_integration.py`.

## Order

Independent of every other modernization issue. Run it in parallel.

## Related

Issue #250 reports that the unit tests are not successful. This work can close that report. Check it.

### PR DESCRIPTION
**Pull request by jaysonsantos**
## What changed

`test/conftest.py` is rewritten. Every fixture now checks that the process
started, waits for the port with a retry loop, and uses `@pytest.fixture` in
place of the deprecated `pytest.yield_fixture`. The unix socket file is removed
before start and on teardown. The IPv6 server moves to port 5002. A new
`memcached_sasl` fixture starts memcached with SASL authentication enabled.

Three test files change or arrive:

- `test/test_ipv6.py` is new. It does a real `set`, `get`, `set_multi`, and
  `get_multi` over IPv6.
- `test/test_sasl_integration.py` is new. It authenticates against a real
  SASL-enabled server and does a `set` and a `get`.
- `test/test_distributed_client_hashing.py` gains four tests. They check the
  key-to-server mapping after a server joins and after a server leaves.

`flake.nix` builds memcached with SASL, in the same way it already builds it
with TLS. It adds `cyrus_sasl.bin` to the dev shell for `saslpasswd2`.

## Why

The fixtures started real memcached processes and never checked the result. A
missing `memcached` binary failed every test with an opaque
`ConnectionRefusedError`. A fixed `time.sleep(0.1)` made a slow start flaky. A
run that died before teardown left `/tmp/memcached.sock` behind, and the next
run then failed to bind.

Two coverage gaps were larger. The IPv6 fixture ran on every session and no
test connected to it, so nothing proved the IPv6 path worked end to end.
`test/test_auth.py` mocks `Protocol._get_response` and feeds canned bytes, so
no test ever ran against a server with authentication turned on.
`test/test_distributed_client_hashing.py` held 12 lines and asserted one key
against one server ten times. It never tested ring rebalance, which is the
property consistent hashing exists for.

## Verification

```
nix develop --command bash -c 'pytest -q'
```

Result: `271 passed`. The baseline on `main` is `261 passed`. This pull request
adds 10 tests and changes no existing count. It also removes four
`PytestDeprecationWarning` lines; one remains, from `test/test_tls.py`.

```
nix develop --command bash -c 'flake8'
```

Result: 0 errors.

The new tests, run alone:

```
nix develop --command bash -c 'pytest -q test/test_ipv6.py test/test_sasl_integration.py'
nix develop --command bash -c 'pytest -q test/test_distributed_client_hashing.py'
```

Result: `6 passed` and `5 passed`.

The skip path, with `memcached` removed from `PATH`:

```
PATH=<without the memcached directory> pytest -q test/test_ipv6.py -rs
```

Result: `3 skipped`, each with
`Cannot run memcached: [Errno 2] No such file or directory: 'memcached'. Is memcached on PATH?`
On `main` this run fails instead, with `ConnectionRefusedError`.

Stale socket recovery:

```
touch /tmp/memcached.sock
nix develop --command bash -c 'pytest -q test/test_socket.py'
```

Result: `44 passed`, and `/tmp/memcached.sock` is gone after the run. On `main`
the memcached process fails to bind.

## Risks

A reviewer must check four points.

1. **The SASL test skips in CI as things stand.** The fixture needs
   `saslpasswd2`, which Ubuntu ships in the `sasl2-bin` package, and a
   memcached built with SASL support. `.ci-before-script.sh` installs neither.
   Adding them belongs to the CI workflow issue (#271), not here. The fixture
   skips cleanly with a clear reason rather than failing. It does run and pass
   in the Nix dev shell.
2. **`flake.nix` changes the memcached build.** It adds `--enable-sasl` and
   `--enable-sasl-pwdb` next to the existing `--enable-tls`, and it adds
   `cyrus_sasl` as a build input. This forces a memcached rebuild on the first
   `nix develop` after checkout. Verified locally: `memcached --help` now lists
   `-S, --enable-sasl`, and the full suite still passes.
3. **The IPv6 fixture moved from port 11211 to port 5002.** It had to. On Linux
   a plain `memcached` binds both `INADDR_ANY` and `IN6ADDR_ANY`, so the
   `memcached_standard_port` fixture already holds 11211 and the `-l::1`
   process cannot bind it. That failure was silent before, because nothing
   checked the process and no test used the fixture.
4. **Fixed ports are still fixed.** Problem 2 in the issue describes port
   collisions between two runners on one host. This pull request does not make
   the ports dynamic. That is not in the acceptance criteria, and the test
   files hardcode `:11211` and `:5000` in many places. What did change is the
   failure mode: a collision now gives one clear skip that names the port and
   the memcached stderr, in place of an opaque error in every test.

Two findings came out of this work. Neither is folded into this pull request.

- **#283** — memcached answers a failed SASL auth with status `0x20`. The
  client maps only `0x08` to `InvalidCredentials`, so a wrong password raises
  the base `MemcachedException`. The mocked test in `test/test_auth.py` hid
  this, because it feeds `0x08` by hand.
  `test/test_sasl_integration.py::test_wrong_password_is_rejected` asserts the
  real behavior and names the issue in its docstring.
- **#284** — `test/test_tls.py:16` still uses the deprecated
  `pytest.yield_fixture`. It is outside this issue's file list.

On the related report: #250 says the unit tests do not run successfully on
Ubuntu. This pull request turns the most likely cause, a missing or slow
`memcached`, into a clear skip message rather than a wall of connection
errors. It does not confirm the original diagnosis, so I have not closed that
issue.

Closes #274

## Summary by CodeRabbit

* **New Features**
  * Added development support for TLS and SASL-enabled memcached environments.
  * Added integration coverage for authenticated memcached connections, including successful and failed authentication.
  * Added IPv6 integration coverage for single-key and multi-key operations.

* **Bug Fixes**
  * Improved test server startup checks and diagnostics.
  * Improved handling of active and stale Unix socket files.
  * Expanded distributed hashing validation for consistent key placement when servers change.

### CODE REVIEW

**Comment by jaysonsantos** [comment]
@claude review

File: test/conftest.py [hunk 1]
```diff
@@ -1,51 +1,221 @@
...
+    deadline = time.time() + START_TIMEOUT
+    last_error = None
+
+    while time.time() < deadline:
+        reason = _fail_reason(process, description)
+        if reason is not None:
+            pytest.skip(reason)
+
+        try:
+            connect().close()
```
[the remark this answers is not in the document]
Comment (jaysonsantos) [reply test/conftest.py:65]: Fixed in the latest push. Each fixture connects to its endpoint before it starts memcached. If a process already listens there, the fixture fails with a clear message, so the tests never use a foreign server. `memcached_socket` no longer yields `None` for an active socket. After a connect succeeds, `_wait_until_accepting` also confirms that the child did not exit. I kept the fixed endpoints because many tests hardcode ports 11211 and 5000 and the socket path.

File: test/conftest.py [hunk 2]
```diff
@@ -1,51 +1,243 @@
...
+
+def _start(args, description, connect):
+    """
+    Start a memcached process and wait for it to accept a connection.
+    """
+    _reject_occupied_endpoint(description, connect)
+    try:
+        process = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    except OSError as error:
+        pytest.skip(f"Cannot run {args[0]}: {error}. Is memcached on PATH?")
```
[the remark this answers is not in the document]
Comment (jaysonsantos) [reply test/conftest.py]: Fixed in the latest push. The standard-port, port-5000, and unix socket fixtures now fail when memcached is missing or does not start. Only the optional IPv6 and SASL fixtures still skip.
